<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2024</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Eduardo Moya
- Ayudantes: Nicolás Ojeda, Melanie Peña, Valentina Rojas

### Equipo: **SUPER IMPORTANTE - notebooks sin nombre no serán revisados**

- Nombre de alumno 1: Joaquín Herrera Suárez
- Nombre de alumno 2: Hecmar Taucare Reyes

### **Link de repositorio de GitHub:** [Repositorio](https://github.com/Joaquin-HS/MDS7202)

## Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 6 días de plazo con descuento de 1 punto por día. Entregas Jueves a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [1]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
#%pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

# Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos. El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.

In [2]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://i.imgur.com/44knnIf.png" width="500" height="500">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```

In [3]:
# Se carga el dataset
df_grades = pd.read_json('students_grades.json')

# Se muestran los primeros registros para revisar el contenido
display(df_grades.head())

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
1  standard                    none        124           142           142  
2  standard                    none        156           196           193  
3  standard               completed         80            78            81  
4  standard                    none        131           166           166

In [4]:
# Se muestran los tipos de datos
df_grades.dtypes

names                          object
gender                         object
race/ethnicity                 object
parental level of education    object
lunch                          object
test preparation course        object
math score                     object
reading score                  object
writing score                  object
dtype: object

In [5]:
# Se especifica el tipo de dato de cada columna
df_grades = pd.read_json('students_grades.json', dtype={
    'names': 'string',
    'gender': 'string',
    'race/ethnicity': 'string',
    'parental level of education': 'string',
    'lunch': 'string',
    'test preparation course': 'string',
    'math score': 'int64',
    'reading score': 'int64',
    'writing score': 'int64'
})

# Se muestran los tipos de datos especificados
df_grades.dtypes

names                          string[python]
gender                         string[python]
race/ethnicity                 string[python]
parental level of education    string[python]
lunch                          string[python]
test preparation course        string[python]
math score                             object
reading score                          object
writing score                          object
dtype: object


<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **1000** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **1000** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier registro que posea un valor diferente a esta escala, **debería ser eliminado de la base de datos**.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?

10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [6]:
def exploratory_data_analysis(df):
    # Se fuerzan valores de las columnas numéricas. Valores no convertibles son nulos.
    # Esto permite que algunos de los aspectos solicitados en el enunciado puedan ser realizados.
    # Las columnas se convierten a tipo Int64 para que sean enteros permitiendo valores nulos.
    df["math score"] = pd.to_numeric(df["math score"], errors='coerce').astype('Int64')
    df["reading score"] = pd.to_numeric(df["reading score"], errors='coerce').astype('Int64')
    df["writing score"] = pd.to_numeric(df["writing score"], errors='coerce').astype('Int64')
    
    # 1. Imprimir un mensaje con las dimensiones del DataFrame
    print(f"El dataframe tiene {df.shape[0]} filas y {df.shape[1]} columnas.\n")

    # 2. Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame
    print("Nombres de las columnas:")
    print(df.columns.tolist(), "\n")

    # 3. Mostrar los primeros 5, los últimos 5 elementos y un muestreo aleatorio de 5 elementos
    print("Primeros 5 elementos:")
    display(df.head())
    print("\nÚltimos 5 elementos:")
    display(df.tail())
    print("\nMuestreo aleatorio de 5 elementos:")
    display(df.sample(5))

    # 4. Mostrar una descripción rápida de las variables numéricas del DataFrame
    print("\nDescripción de las variables numéricas:")
    display(df.describe())

    # 5. Imprimir un mensaje con la cantidad de valores nulos en el DataFrame
    print("\nCantidad de valores nulos:")
    print(df.isna().sum())

    # 6. Imprimir un mensaje que detalle la cantidad de valores únicos en el DataFrame
    print("\nCantidad de valores únicos:")
    print(df.nunique())

    # 7. Imprimir un mensaje con el total de filas que contengan nombres duplicados en el DataFrame
    print("\nCantidad de nombres duplicados:")
    print(df.duplicated(subset='names').sum())

    # 8. Mostrar las filas que contengan nombres sean duplicados en el DataFrame
    print("\nFilas con nombres duplicados:")
    display(df[df.duplicated(subset='names', keep="first")])  # Se mantiene la primera ocurrencia

In [7]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

In [8]:
exploratory_data_analysis(df_grades)

El dataframe tiene 1400 filas y 9 columnas.

Nombres de las columnas:
['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score'] 

Primeros 5 elementos:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none          37             46             46  
1  standard                    none         124            142            142  
2  standard                    none         156            196            193  
3  standard               completed          80             78             81  
4  standard                    none         131            166            166


Últimos 5 elementos:


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1395      standard                    none         193            142   
1396  free/reduced               completed          66             74   
1397      standard               completed          54             61   
1398  free/reduced                    none         193            196   
1399      standard               completed          68             80   

      writing score  
1395            193  
1396             81  
1397             62  
1398            193  
1399             76


Muestreo aleatorio de 5 elementos:


names  gender race/ethnicity parental level of education  \
281      Georgia Collins  female        group B            some high school   
956     Jennie Blackwood  female        group C          associate's degree   
171        Judith Foster  female        group C           bachelor's degree   
1357     Jeffrey Bostrom    male        group E          associate's degree   
1178  Christopher Emmons    male        group C                 high school   

             lunch test preparation course  math score  reading score  \
281       standard                    none          62             64   
956       standard                    none          65             76   
171       standard                    none          83             93   
1357      standard               completed          71             74   
1178  free/reduced               completed          58             51   

      writing score  
281              66  
956              76  
171              95  
1357             68  
1178             52


Descripción de las variables numéricas:


math score  reading score  writing score
count      1360.0         1360.0         1353.0
mean    89.538971      93.086029      91.852919
std     47.152422      46.385775       47.04357
min           0.0           17.0           10.0
25%          59.0           63.0           62.0
50%          72.0           76.0           75.0
75%         97.25          100.0          100.0
max         196.0          196.0          196.0


Cantidad de valores nulos:
names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64

Cantidad de valores únicos:
names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64

Cantidad de nombres duplicados:
247

Filas con nombres duplicados:


names  gender race/ethnicity parental level of education  \
12         Rita Courtney  female        group B            some high school   
81    Elizabeth Quintero  female        group B                some college   
160    Michelle Gonzales  female        group B            some high school   
170        Carla Mendoza  female        group D          associate's degree   
176       Darlene Parker  female        group C          associate's degree   
...                  ...     ...            ...                         ...   
1385      James Crawford    male        group D             master's degree   
1390     Gloria Bradford  female        group B                 high school   
1393         Adam Romero    male        group C                 high school   
1395    Alexander Dillon    male        group D          associate's degree   
1398        Joseph Mccoy    male        group D                some college   

             lunch test preparation course  math score  reading score  \
12        standard                    none         193            193   
81        standard                    none          82             85   
160       standard               completed          60             70   
170   free/reduced                    none          43             60   
176       standard                    none          54             61   
...            ...                     ...         ...            ...   
1385      standard                    none         193            147   
1390      standard                    none         124             64   
1393      standard               completed         124            193   
1395      standard                    none         193            142   
1398  free/reduced                    none         193            196   

      writing score  
12              193  
81               87  
160              74  
170              58  
176              58  
...             ...  
1385            131  
1390            193  
1393             49  
1395            193  
1398            193  

[247 rows x 9 columns]

**Justificación:**

En primer lugar, se destaca que existen valores nulos dentro de las columnas numéricas del dataframe, lo que quiere decir que algunas personas no cuentan con alguna o algunas calificaciones. Por otro lado, existen varias filas duplicadas, específicamente 247. Es importante también destacar que existen 1400 filas, cuando deberían ser 1000 según lo especificado por el rector. Si se descuentan las 247 duplicadas, quedan 1153 en total (mismo valor de nombres únicos), con 153 pendientes de revisión.

Analizando los *boxplots* generados (y la descripción de las variables numéricas) se aprecia que, para cada evaluación, aproximadamente un 25% de las notas está sobre la calificación máxima posible (100), lo que genera una distorsión en los gráficos. Siguiendo lo mencionado por el rector, se deben eliminar de la base de datos dichos registros.

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y **cree una función** que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**
- [ ] Eliminar registros con valores nulos.
- [ ] Eliminar valores fuera del rango válido.
- [ ] Eliminar registros con nombres duplicados
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hints**

- Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.
- Cuidado con el orden en la limpieza de datos. ¿Que se debe eliminar primero, los nombres duplicados o los registros fuera del rango posible?

> Utilice la función creada en el punto 1 para verificar sus resultados.

**Respuesta:**

In [9]:
def clean_data(df):
    print(f"El dataframe tiene {df.shape[0]} filas y {df.shape[1]} columnas.\n")

    # En primer lugar se eliminan registros con nombres duplicados, ya que no es posible que un alumno haya
    # dado más de una vez el mismo examen. Así, se está trabajando solo con registros válidos.
    print("Se eliminan registros duplicados.")
    df = df.drop_duplicates(subset=["names"], keep="first")
    print(f"El dataframe tiene {df.shape[0]} filas y {df.shape[1]} columnas.\n")

    # El segundo paso sería eliminar registros con valores nulos (de haber), ya que se está trabajando con
    # registros válidos pero por alguna razón algunas celdas no están completas.
    print("Se eliminan registros con valores nulos.")
    df.dropna(inplace=True)
    print(f"El dataframe tiene {df.shape[0]} filas y {df.shape[1]} columnas.\n")

    # Por último, teniendo registros válidos y completos, se filtran para dejar solo aquellas filas
    # que poseen valores dentro del rango admitido de notas (entre 0 y 100).
    print("Se filtran las filas con valores dentro del rango admitido de notas (entre 0 y 100).")
    df = df[
        (df["math score"] >= 0) & (df["math score"] <= 100)
        & (df["reading score"] >= 0) & (df["reading score"] <= 100)
        & (df["writing score"] >= 0) & (df["writing score"] <= 100)
    ]
    print(f"El dataframe tiene {df.shape[0]} filas y {df.shape[1]} columnas.\n")

    return df

In [10]:
df_grades = clean_data(df_grades)
df_grades

El dataframe tiene 1400 filas y 9 columnas.

Se eliminan registros duplicados.
El dataframe tiene 1153 filas y 9 columnas.

Se eliminan registros con valores nulos.
El dataframe tiene 1084 filas y 9 columnas.

Se filtran las filas con valores dentro del rango admitido de notas (entre 0 y 100).
El dataframe tiene 875 filas y 9 columnas.



C:\Users\joaco\AppData\Local\Temp\ipykernel_24676\2523818197.py:13: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none          37             46   
3         standard               completed          80             78   
5         standard                    none          73             66   
6         standard               completed          81             75   
7         standard               completed          94             86   
...            ...                     ...         ...            ...   
1392      standard                    none          69             75   
1394  free/reduced               completed          23             44   
1396  free/reduced               completed          66             74   
1397      standard               completed          54             61   
1399      standard               completed          68             80   

      writing score  
0                46  
3                81  
5                63  
6                76  
7                87  
...             ...  
1392             71  
1394             36  
1396             81  
1397             62  
1399             76  

[875 rows x 9 columns]

In [11]:
exploratory_data_analysis(df_grades)

El dataframe tiene 875 filas y 9 columnas.

Nombres de las columnas:
['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score'] 

Primeros 5 elementos:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none          37             46             46  
3  standard               completed          80             78             81  
5  standard                    none          73             66             63  
6  standard               completed          81             75             76  
7  standard               completed          94             86             87


Últimos 5 elementos:


names  gender race/ethnicity parental level of education  \
1392   Richard Young    male        group D                 high school   
1394   Wanda Russell  female        group B                 high school   
1396  Marina Zeigler  female        group C           bachelor's degree   
1397   Laurie Carter  female        group B            some high school   
1399    Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1392      standard                    none          69             75   
1394  free/reduced               completed          23             44   
1396  free/reduced               completed          66             74   
1397      standard               completed          54             61   
1399      standard               completed          68             80   

      writing score  
1392             71  
1394             36  
1396             81  
1397             62  
1399             76


Muestreo aleatorio de 5 elementos:


names  gender race/ethnicity parental level of education  \
474      Randy Mccarthy    male        group E          associate's degree   
30        Janine Taylor  female        group C          associate's degree   
1150  Kimberly Routzahn  female        group C             master's degree   
628          Mary Kovac  female        group C          associate's degree   
1281      Crystal Smith  female        group C                 high school   

             lunch test preparation course  math score  reading score  \
474       standard                    none          72             57   
30    free/reduced                    none          64             73   
1150  free/reduced                    none          52             65   
628       standard               completed          68             67   
1281  free/reduced                    none          66             76   

      writing score  
474              62  
30               68  
1150             61  
628              73  
1281             68


Descripción de las variables numéricas:


math score  reading score  writing score
count       875.0          875.0          875.0
mean    66.225143      69.195429      68.114286
std     15.076716       14.62927       15.17422
min           0.0           17.0           10.0
25%          57.0           59.0           57.5
50%          66.0           70.0           69.0
75%          76.5           79.0           79.0
max         100.0          100.0          100.0


Cantidad de valores nulos:
names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

Cantidad de valores únicos:
names                          875
gender                           2
race/ethnicity                   5
parental level of education      6
lunch                            2
test preparation course          2
math score                      79
reading score                   72
writing score                   77
dtype: int64

Cantidad de nombres duplicados:
0

Filas con nombres duplicados:


Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

**Justificación:**

En primer lugar se eliminan registros con nombres duplicados, ya que no es posible que un alumno haya dado más de una vez el mismo examen. Así, se está trabajando solo con registros válidos.

El segundo paso sería eliminar registros con valores nulos (de haber), ya que se está trabajando con registros válidos pero por alguna razón algunas celdas no están completas.

Por último, teniendo registros válidos y completos, se filtran para dejar solo aquellas filas que poseen valores dentro del rango admitido de notas (entre 0 y 100).

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [12]:
def nota_chilena(dataframe_in, columna):
    dataframe_in[columna] = dataframe_in[columna].apply(lambda x: round(x*0.06 + 1, 2))  # "x" es el % de logro
    # No se retorna nada, ya que se modifica el dataframe directamente

In [13]:
nota_chilena(df_grades, "math score")
nota_chilena(df_grades, "reading score")
nota_chilena(df_grades, "writing score")
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score  
0              3.76  
3              5.86  
5              4.78  
6              5.56  
7              6.22  
...             ...  
1392           5.26  
1394           3.16  
1396           5.86  
1397           4.72  
1399           5.56  

[875 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [14]:
def alumnos_promedio(dataframe_in):
    # Notas redondeadas al segundo decimal
    dataframe_in[["math score", "reading score", "writing score"]] = dataframe_in[["math score", "reading score", "writing score"]].round(2)

    # Se calcula el promedio de las tres notas, redondeado al segundo decimal
    dataframe_in["GPA"] = dataframe_in[["math score", "reading score", "writing score"]].mean(axis=1).round(2)

    return dataframe_in  # Se podría no haber retornado igual que antes. Este enfoque es distinto pero funciona igual.

In [15]:
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
3  standard               completed        5.80           5.68           5.86   
5  standard                    none        5.38           4.96           4.78   
6  standard               completed        5.86           5.50           5.56   
7  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
3  5.78  
5  5.04  
6  5.64  
7  6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [16]:
def reprobados_destacados(dataframe_in):
    # Se identifican los alumnos reprobados (promedio de notas inferior a 4.0):
    dataframe1 = dataframe_in[dataframe_in["GPA"] < 4.0]

    # Se entrega un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados:
    dataframe2 = dataframe_in.copy()
    dataframe2["GPA"] = dataframe2["GPA"].apply(lambda x: "R" if x < 4.0 else x)

    # Se identifican los alumnos destacados (notas superiores a 5.8 en todas las asignaturas) ordenados de mayor a menor:
    dataframe3 = dataframe_in[
        (dataframe_in["math score"] > 5.8)
        & (dataframe_in["reading score"] > 5.8)
        & (dataframe_in["writing score"] > 5.8)
    ].sort_values(by="GPA", ascending=False)

    return dataframe1, dataframe2, dataframe3

In [17]:
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)

print("Alumnos reprobados (promedio de notas inferior a 4.0):")
display(dataframe1)
print("\nAlumnos marcando con R el GPA para alumnos reprobados:")
display(dataframe2)
print("\nAlumnos destacados (notas superiores a 5.8 en todas las asignaturas) ordenados de mayor a menor:")
display(dataframe3)

Alumnos reprobados (promedio de notas inferior a 4.0):


names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
13        Mikki Amaya  female        group A                 high school   
26      Samuel Dougal    male        group C             master's degree   
55    Rodger Talavera    male        group B                 high school   
63     Morris Johnson    male        group A          associate's degree   
...               ...     ...            ...                         ...   
1371   Charles Harian    male        group C            some high school   
1373    Wendy Gunnell  female        group D            some high school   
1389      Diana Costa  female        group C                 high school   
1391      Amy Sieving  female        group C                 high school   
1394    Wanda Russell  female        group B                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
13    free/reduced               completed        3.04           3.88   
26    free/reduced               completed        3.76           3.52   
55    free/reduced                    none        3.16           2.74   
63    free/reduced               completed        3.40           4.30   
...            ...                     ...         ...            ...   
1371  free/reduced               completed        3.70           4.12   
1373  free/reduced                    none        2.62           3.04   
1389  free/reduced                    none        3.46           3.76   
1391      standard                    none        2.74           2.74   
1394  free/reduced               completed        2.38           3.64   

      writing score   GPA  
0              3.76  3.58  
13             3.46  3.46  
26             3.76  3.68  
55             2.62  2.84  
63             4.18  3.96  
...             ...   ...  
1371           3.94  3.92  
1373           2.92  2.86  
1389           3.58  3.60  
1391           2.80  2.76  
1394           3.16  3.06  

[86 rows x 10 columns]


Alumnos marcando con R el GPA para alumnos reprobados:


names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score   GPA  
0              3.76     R  
3              5.86  5.78  
5              4.78  5.04  
6              5.56  5.64  
7              6.22  6.34  
...             ...   ...  
1392           5.26   5.3  
1394           3.16     R  
1396           5.86  5.42  
1397           4.72  4.54  
1399           5.56  5.48  

[875 rows x 10 columns]


Alumnos destacados (notas superiores a 5.8 en todas las asignaturas) ordenados de mayor a menor:


names  gender race/ethnicity parental level of education  \
384   Elizabeth Beasley  female        group E           bachelor's degree   
677       Richard Pryor    male        group E           bachelor's degree   
1352      Sandra Pompey  female        group E          associate's degree   
475           Emma Gray  female        group E           bachelor's degree   
1276     Angela Cressey  female        group D                some college   
...                 ...     ...            ...                         ...   
949      Rolando Widmer    male        group C                 high school   
296   Lucille Stanphill  female        group C                 high school   
300        Steven Kelly    male        group D                some college   
1307        Peter Perez    male        group D             master's degree   
96          Harry Berry    male        group B          associate's degree   

         lunch test preparation course  math score  reading score  \
384   standard                    none        7.00           7.00   
677   standard               completed        7.00           7.00   
1352  standard                    none        7.00           7.00   
475   standard               completed        6.94           7.00   
1276  standard                    none        6.88           7.00   
...        ...                     ...         ...            ...   
949   standard               completed        5.92           6.04   
296   standard                    none        5.86           6.04   
300   standard                    none        5.86           5.92   
1307  standard                    none        5.86           5.86   
96    standard               completed        5.86           5.92   

      writing score   GPA  
384            7.00  7.00  
677            7.00  7.00  
1352           7.00  7.00  
475            7.00  6.98  
1276           6.94  6.94  
...             ...   ...  
949            5.92  5.96  
296            5.92  5.94  
300            6.04  5.94  
1307           6.04  5.92  
96             5.92  5.90  

[94 rows x 10 columns]

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [18]:
# Se genera una nueva columna que clasifica en quintiles de rendimiento los alumnos: ['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']
df_grades["GPA cuantiles"] = pd.qcut(df_grades["GPA"], q=5, labels=['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto'])

In [19]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="GPA"),
    x="GPA",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    #color_continuous_scale="magma"
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig

A partir del gráfico se pueden extraer algunos fenómenos interesantes:
- A medida que aumenta el nivel educacional de los padres los promedios "muy altos" tienden a ser mayores, hasta que se llega a nivel de "college" o algún "degree" donde se estabiliza más la distribución de dichos promedios. Por su parte, siguiendo la misma tendencia de nivel educacional en los padres, los promedios "bajísimos" tienden a aumentar, generando que exista un comportamiento menos disparejo en esta categoría (menor número de outliers)
- Los cuantiles asociados a promedios bajos, medios y altos tienen comportamientos aproximadamente invariables al nivel educacional de los padres.

Con esto, es en los casos borde (promedios bajísimos o muy altos) donde el nivel de educación de los padres pasa a ser relevante sobre el promedio de notas de los estudiantes.

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

**Respuesta:**

In [20]:
df_grades_1 = pd.read_json('students_grades_1.json')
df_grades_2 = pd.read_json('students_grades_2.json')

df_grades = pd.concat([df_grades_1, df_grades_2], axis=0)

In [21]:
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
998      Richard Young    male        group D                 high school   
999      Wanda Russell  female        group B                 high school   
1000    Marina Zeigler  female        group C           bachelor's degree   
1001     Laurie Carter  female        group B            some high school   
1002      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
1         standard               completed        5.80           5.68   
2         standard                    none        5.38           4.96   
3         standard               completed        5.86           5.50   
4         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
998       standard                    none        5.14           5.50   
999   free/reduced               completed        2.38           3.64   
1000  free/reduced               completed        4.96           5.44   
1001      standard               completed        4.24           4.66   
1002      standard               completed        5.08           5.80   

      writing score   GPA GPA cuantiles  
0              3.76  3.58      Bajísimo  
1              5.86  5.78          Alto  
2              4.78  5.04         Medio  
3              5.56  5.64          Alto  
4              6.22  6.34      Muy alto  
...             ...   ...           ...  
998            5.26  5.30         Medio  
999            3.16  3.06      Bajísimo  
1000           5.86  5.42          Alto  
1001           4.72  4.54          Bajo  
1002           5.56  5.48          Alto  

[1000 rows x 11 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos]**


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


**Respuesta:**

In [22]:
def tabla_resumen(df):
    # Se genera una tabla resumen con lo solicitado

    # Se calcula el promedio de las notas de cada asignatura
    df["GPA"] = df[["math score", "reading score", "writing score"]].mean(axis=1).round(2)

    # Se mapea la variable "parental level of education":
    df["simple parental level of education"] = df["parental level of education"].map({
        "some high school": "school",
        "some college": "school",
        "high school": "school",
        "bachelor's degree": "college",
        "associate's degree": "college",
        "master's degree": "postgraduate"
    })

    # Se agrega según "race/ethnicity" y "simple parental level of education" obteniendo el promedio de notas
    prom = df.groupby(["race/ethnicity", "simple parental level of education"])[["math score", "reading score", "writing score", "GPA"]].mean().round(2)

    # Se agrega según "race/ethnicity" y "simple parental level of education" obteniendo el conteo de alumnos
    cont = df.groupby(["race/ethnicity", "simple parental level of education"]).size().reset_index(name='count')

    # Se unen ambos DataFrames generando la tabla resumen
    resumen = pd.merge(prom, cont, on=["race/ethnicity", "simple parental level of education"])
           
    # Se obtiene el porcentaje de alumnos con respecto al total
    total_alumnos = df.shape[0]
    resumen['percentage'] = (resumen['count'] / total_alumnos * 100).round(2).astype(str) + '%'

    return resumen

In [23]:
resumen = tabla_resumen(df_grades)
resumen

race/ethnicity simple parental level of education  math score  \
0         group A                            college        4.83   
1         group A                       postgraduate        4.46   
2         group A                             school        4.65   
3         group B                            college        5.03   
4         group B                       postgraduate        5.03   
5         group B                             school        4.69   
6         group C                            college        5.03   
7         group C                       postgraduate        5.02   
8         group C                             school        4.74   
9         group D                            college        5.06   
10        group D                       postgraduate        5.35   
11        group D                             school        4.99   
12        group E                            college        5.53   
13        group E                       postgraduate        5.48   
14        group E                             school        5.35   

    reading score  writing score   GPA  count percentage  
0            5.05           4.95  4.94     26       2.6%  
1            4.88           5.06  4.80      3       0.3%  
2            4.81           4.66  4.71     60       6.0%  
3            5.24           5.16  5.14     61       6.1%  
4            5.81           5.63  5.49      6       0.6%  
5            4.90           4.79  4.79    123      12.3%  
6            5.36           5.33  5.24    118      11.8%  
7            5.23           5.17  5.14     19       1.9%  
8            5.00           4.89  4.88    182      18.2%  
9            5.22           5.24  5.17     78       7.8%  
10           5.63           5.78  5.59     23       2.3%  
11           5.13           5.11  5.08    161      16.1%  
12           5.45           5.43  5.47     57       5.7%  
13           5.93           5.83  5.74      8       0.8%  
14           5.27           5.11  5.25     75       7.5%

In [28]:
fig = px.box(
    df_grades,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
    category_orders={ # Se agrega esto para que aparezcan ordenadas las categorías
        "race/ethnicity": ["group A", "group B", "group C", "group D", "group E"],
        "simple parental level of education": ["school", "college", "postgraduate"]
    }
)
fig.show()

Se mantiene el comportamiento general mencionado previamente con respecto a un aumento del GPA a medida que crece el nivel de educación de los padres. Se aprecia, además, que existe una menor variabilidad en el promedio de notas en el caso de que los padres tengan un nivel de educación "postgraduate", comparando con el caso "school" y "college" donde las distribuciones de GPA son más dispersas.

Con respecto a la etnicidad de los padres se puede decir que los grupos B y E cuentan con mejores valores de GPA en general, mientras que los grupos A, C y D tienden a tener promedios más bajos y con mayor variabilidad. Se destaca que estos fenómenos se hacen menos apreciables a medida que disminuye el nivel educacional de los padres, debido a la existencia de distribuciones más dispersas como se dijo previamente. Por último, en esa misma tendencia se destaca la aparición de un mayor número de outliers.

**Resultados Esperados**

Esta tabla ilustra la estructura del dataframe a conseguir. Los valores son referenciales, no necesariamente son los esperados.

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.83 |            5.05 |            4.95 |  4.94 |      26 | 2.6 %        |
|  1 |                  | postgraduate                         |         4.46 |            4.88 |            5.06 |  4.80 |       3 | 0.3 %        |
|  2 |                  | school                               |         4.65 |            4.81 |            4.66 |  4.71 |      60 | 6.0 %        |
|  3 | group B          | college                              |         5.03 |            5.24 |            5.16 |  5.14 |      61 | 6.1 %        |
|  4 |                  | postgraduate                         |         5.03 |            5.81 |            5.63 |  5.49 |       6 | 0.6 %        |
|  5 |                  | school                               |         4.69 |            4.90 |            4.79 |  4.79 |     123 | 12.3 %       |
|  6 | group C          | college                              |         5.03 |            5.36 |            5.33 |  5.24 |     118 | 11.8 %       |
|  7 |                  | postgraduate                         |         5.02 |            5.23 |            5.17 |  5.14 |      19 | 1.9 %        |
|  8 |                  | school                               |         4.74 |            5.00 |            4.89 |  4.88 |     182 | 18.2 %       |
|  9 | group D          | college                              |         5.06 |            5.22 |            5.24 |  5.17 |      78 | 7.8 %        |
| 10 |                  | postgraduate                         |         5.35 |            5.63 |            5.78 |  5.59 |      23 | 2.3 %        |
| 11 |                  | school                               |         4.99 |            5.13 |            5.11 |  5.08 |     161 | 16.1 %       |
| 12 | group E          | college                              |         5.53 |            5.45 |            5.43 |  5.47 |      57 | 5.7 %        |
| 13 |                  | postgraduate                         |         5.48 |            5.93 |            5.83 |  5.74 |       8 | 0.8 %        |
| 14 |                  | school                               |         5.35 |            5.27 |            5.11 |  5.25 |      75 | 7.5 %        |


### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene registros de alumnos que ya no forman parte de la universidad, por lo que no deben ser parte de este análisis.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados o **alumnos que no pertenecen a la universidad**).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

**Respuesta:**

In [26]:
# Se cargan los archivos
students = pd.read_json('students_grades.json')
other_grades = pd.read_csv('other_grades.csv')

# Outer join (se incluyen todos los estudiantes, independiente de si están en ambas tablas)
df_grades_outer = students.merge(other_grades, how='outer', left_on='names', right_on='names')

# Left join (se incluyen todos los estudiantes del DataFrame original)
df_grades_left = students.merge(other_grades, how='left', left_on='names', right_on='names')

# Right join (se incluyen todos los estudiantes del DataFrame de otras notas)
df_grades_right = students.merge(other_grades, how='right', left_on='names', right_on='names')

# Inner join (se incluyen solo los estudiantes que están en ambos DataFrames)
df_grades_inner = students.merge(other_grades, how='inner', left_on='names', right_on='names')

In [27]:
print("Outer join:", df_grades_outer.shape)
display(df_grades_outer)
print("Left join:", df_grades_left.shape)
display(df_grades_left)
print("Right join:", df_grades_right.shape)
display(df_grades_right)
print("Inner join:", df_grades_inner.shape)
display(df_grades_inner)

Outer join: (1550, 11)


names gender race/ethnicity parental level of education  \
0       Aaron Morgan    NaN            NaN                         NaN   
1       Aaron Oakden   male        group D                 high school   
2       Aaron Oakden   male        group D                 high school   
3       Aaron Shelly   male        group D          associate's degree   
4       Abigail Hall    NaN            NaN                         NaN   
...              ...    ...            ...                         ...   
1545   Zachary Glenn   male        group E                some college   
1546  Zachary Torres    NaN            NaN                         NaN   
1547   Zoey Campbell    NaN            NaN                         NaN   
1548       Zoey Long    NaN            NaN                         NaN   
1549  Zoey Patterson    NaN            NaN                         NaN   

             lunch test preparation course math score reading score  \
0              NaN                     NaN        NaN           NaN   
1     free/reduced                    none         66            74   
2     free/reduced                    none                      193   
3         standard                    none         61            48   
4              NaN                     NaN        NaN           NaN   
...            ...                     ...        ...           ...   
1545      standard               completed         99            87   
1546           NaN                     NaN        NaN           NaN   
1547           NaN                     NaN        NaN           NaN   
1548           NaN                     NaN        NaN           NaN   
1549           NaN                     NaN        NaN           NaN   

     writing score  science score  history score  
0              NaN           3.35           5.46  
1               69           3.76           4.60  
2              124           3.76           4.60  
3               46           3.88           3.90  
4              NaN           2.88           4.59  
...            ...            ...            ...  
1545            81           6.94           5.70  
1546           NaN           4.48           6.92  
1547           NaN           4.05           1.72  
1548           NaN           1.75           3.50  
1549           NaN           4.89           6.68  

[1550 rows x 11 columns]

Left join: (1400, 11)


names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1      Madeline Fuller  female        group C            some high school   
2        Daniel Rester    male        group E          associate's degree   
3     Charles Linstrom    male        group A           bachelor's degree   
4       Geraldine Hale  female        group A          associate's degree   
...                ...     ...            ...                         ...   
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0         standard                    none         37            46   
1         standard                    none        124           142   
2         standard                    none        156           196   
3         standard               completed         80            78   
4         standard                    none        131           166   
...            ...                     ...        ...           ...   
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  science score  history score  
0               46           2.50            3.8  
1              142            NaN            NaN  
2              193            NaN            NaN  
3               81           4.36            5.2  
4              166            NaN            NaN  
...            ...            ...            ...  
1395           193           5.80            5.0  
1396            81           4.54            4.1  
1397            62           3.58            4.3  
1398           193           4.30            3.4  
1399            76           4.66            5.3  

[1400 rows x 11 columns]

Right join: (1397, 11)


names  gender race/ethnicity parental level of education  \
0           Pam Otoole  female        group B           bachelor's degree   
1          Diane Olson  female        group C                some college   
2       Sarah Flannery  female        group B             master's degree   
3     Stanley Chappell    male        group A          associate's degree   
4        Arnold Dodson    male        group C                some college   
...                ...     ...            ...                         ...   
1392     Carson Brooks     NaN            NaN                         NaN   
1393         Ruby Reed     NaN            NaN                         NaN   
1394    Zachary Torres     NaN            NaN                         NaN   
1395       Eliana Gray     NaN            NaN                         NaN   
1396       Gavin James     NaN            NaN                         NaN   

             lunch test preparation course math score reading score  \
0         standard                    none         72            72   
1         standard               completed         69            90   
2         standard                    none         90            95   
3     free/reduced                    none         47            57   
4         standard                    none         76            78   
...            ...                     ...        ...           ...   
1392           NaN                     NaN        NaN           NaN   
1393           NaN                     NaN        NaN           NaN   
1394           NaN                     NaN        NaN           NaN   
1395           NaN                     NaN        NaN           NaN   
1396           NaN                     NaN        NaN           NaN   

     writing score  science score  history score  
0               74           4.42           4.00  
1               88           3.88           4.80  
2               93           5.32           6.10  
3               44           3.82           3.70  
4               75           4.60           5.20  
...            ...            ...            ...  
1392           NaN           4.72           1.54  
1393           NaN           1.71           4.59  
1394           NaN           4.48           6.92  
1395           NaN           6.91           5.41  
1396           NaN           6.60           6.06  

[1397 rows x 11 columns]

Inner join: (1247, 11)


names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1242  Alexander Dillon    male        group D          associate's degree   
1243    Marina Zeigler  female        group C           bachelor's degree   
1244     Laurie Carter  female        group B            some high school   
1245      Joseph Mccoy    male        group D                some college   
1246      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0         standard                    none         37            46   
1         standard               completed         80            78   
2         standard                    none         73            66   
3         standard               completed         81            75   
4         standard               completed         94            86   
...            ...                     ...        ...           ...   
1242      standard                    none        193           142   
1243  free/reduced               completed         66            74   
1244      standard               completed         54            61   
1245  free/reduced                    none        193           196   
1246      standard               completed         68            80   

     writing score  science score  history score  
0               46           2.50            3.8  
1               81           4.36            5.2  
2               63           4.48            3.8  
3               76           5.32            5.0  
4               87           6.64            4.6  
...            ...            ...            ...  
1242           193           5.80            5.0  
1243            81           4.54            4.1  
1244            62           3.58            4.3  
1245           193           4.30            3.4  
1246            76           4.66            5.3  

[1247 rows x 11 columns]

En primer lugar, dado que hay que considerar `students_grades.json`, se destaca que los registros no están limpios ni en un formato correcto (formato chileno, con duplicados, etc). Obviando este punto, se muestra a continuación un análisis para cada tipo de join:

- Outer join: Este formato de join incluye todos los estudiantes de ambos dataframes, es decir, los que están en ambos y en uno o el otro de manera individual. Esto genera un dataframe con un gran número de registros, donde muchos de ellos contienen nulos.

- Left join: Este formato de join incluye todos los estudiantes del dataframe original, anexando las notas de `science` y `history` solo si hay coincidencias con el nombre.

- Right join: De manera similar al left join, el right join incluye todos los estudiantes del dataframe asociado a las notas de `science` y `history`, agregando la información de `students_grades` solo si hay coincidencias en los nombres.

- Inner join: Este tipo de join incluye solamente a los estudiantes que están en ambos dataframes.

Se considera que el join más adecuado sería el **inner join**, ya que permite asegurar que solo se incluyan los estudiantes que están en ambos dataframes, evitando registros asociados a personas que no están en la universidad (limpiando filas de ambos dataframes). Es importante destacar que esto no necesariamente generaría el dataframe final totalmente limpio, pero ayuda a reducir un gran número de problemas de limpieza que puede tener cada dataframe.

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>